src/pow.rs

```diff
@@ -0,0 +1,1164 @@
+///
+/// Difficulty is a measure of how difficult it is to find a hash below a given target. (See also [`Target`].)
+///
+/// ref: https://en.bitcoin.it/wiki/Difficulty
+///
+/// # Examples
+///
+/// ```
+/// use bitcoin::Difficulty;
+/// let _ = Difficulty::from(0xdeadbeef);
+/// ```
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct Difficulty(Uint256);
+impl_from_and_impl!(Difficulty);
+impl_core_traits!(Difficulty);
+
+impl Difficulty {
+    /// Converts this [`Difficulty`] to a [`Target`].
+    #[inline]
+    pub fn to_target(&self) -> Target { todo!("inverse of Target::to_difficulty()") }
+
+    /// Increments `self` by 1.
+    ///
+    /// # Examples
+    /// ```
+    /// # use bitcoin::Difficulty;
+    /// let mut x = Difficulty::from(1);
+    /// x.increment();
+    /// assert_eq!(x, Difficulty::from(2));
+    /// ```
+    #[inline]
+    pub fn increment(&mut self) { self.0.increment() }
+}
+
+/// A 256 bit integer representing target.
+///
+/// The SHA-256 hash of a block's header must be lower than or equal to the current target for the block to be accepted by the network. The lower the target, the more difficult it is to generate a block. (See also [`Difficulty`].)
+///
+/// ref: https://en.bitcoin.it/wiki/Target
+///
+/// # Examples
+///
+/// ```
+/// use bitcoin::Target;
+/// let _ = Target::from(0xdeadbeef);
+/// ```
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct Target(Uint256);
+impl_from_and_impl!(Target);
+impl_core_traits!(Target);
+
+impl Target {
+    /// Converts this [`Target`] to a [`Difficulty`].
+    pub fn to_difficulty(&self) -> Difficulty {
+        // 2**256 / (target + 1) == ~target / (target+1) + 1    (eqn shamelessly stolen from bitcoind)
+        let mut ret = !self.0;
+        let mut ret1 = self.0;
+        ret1.increment();
+        ret = ret / ret1;
+        ret.increment();
+        Difficulty(ret)
+    }
+
+    /// Computes the [`Target`] value from a compact representation.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use bitcoin::Target;
+    /// assert_eq!(0x1d00ffff, Target::to_compact(&Target::from_compact(0x1d00ffff)));
+    /// ```
+    pub fn from_compact(bits: u32) -> Target {
+        // This is a floating-point "compact" encoding originally used by
+        // OpenSSL, which satoshi put into consensus code, so we're stuck
+        // with it. The exponent needs to have 3 subtracted from it, hence
+        // this goofy decoding code:
+        let (mant, expt) = {
+            let unshifted_expt = bits >> 24;
+            if unshifted_expt <= 3 {
+                ((bits & 0xFFFFFF) >> (8 * (3 - unshifted_expt as usize)), 0)
+            } else {
+                (bits & 0xFFFFFF, 8 * ((bits >> 24) - 3))
+            }
+        };
+
+        // The mantissa is signed but may not be negative
+        if mant > 0x7FFFFF {
+            Target::default()
```

Comment:
Shouldn't we return `Err` instead?

Reply:
Based on this code comment I did not touch this function's code
```
        // This is a floating-point "compact" encoding originally used by
        // OpenSSL, which satoshi put into consensus code, so we're stuck
        // with it. The exponent needs to have 3 subtracted from it, hence
        // this goofy decoding code:
```

Reply:
I found a proper answer,

> When parsing nBits, Bitcoin Core converts a negative target threshold into a target of zero, which the header hash can equal (in theory, at least).

ref: https://developer.bitcoin.org/reference/block_chain.html#target-nbits